src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh

```diff
@@ -0,0 +1,410 @@
+ * \f$ V_{o}, R, a\f$ are the potential well depth, nuclear radius, and
+ * surface thickness (skin depth), respectively.
+ */
+real_type
+NuclearZoneBuilder::integrate_woods_saxon(real_type rmin,
+                                          real_type rmax,
+                                          real_type nuclear_radius) const
+{
+    real_type skin_ratio = nuclear_radius / skin_depth_;
+
+    real_type ws_shift = 2 * skin_ratio;
+    real_type delta_r = rmax - rmin;
+
+    // Average value of \f$ r^{2} V(r) dr \f$ at boundaries
+    real_type integral = half() * delta_r
+                         * (rmin * (rmin + ws_shift) / (1 + std::exp(rmin))
+                            + rmax * (rmax + ws_shift) / (1 + std::exp(rmax)));
+
+    real_type result = integrate_potential(rmin, delta_r, integral, ws_shift);
+
+    return ipow<3>(skin_depth_)
+           * (result
+              + ipow<2>(skin_ratio)
+                    * std::log((1 + std::exp(-rmin)) / (1 + std::exp(-rmax))));
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Integrate the Gaussian potential function from \f$ r_1 \f$ to \f$ r_2 \f$.
+ */
+real_type NuclearZoneBuilder::integrate_gaussian(real_type rmin,
+                                                 real_type rmax,
+                                                 real_type gauss_radius) const
+{
+    real_type delta_r = rmax - rmin;
+    real_type rmin_sq = ipow<2>(rmin);
+    real_type rmax_sq = ipow<2>(rmax);
+
+    real_type integral
+        = this->half() * delta_r
+          * (rmin_sq * std::exp(-rmin_sq) + rmax_sq * std::exp(-rmax_sq));
+
+    real_type result = integrate_potential(false, rmin, delta_r, integral);
+
+    return ipow<3>(gauss_radius) * result;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Integrate a nuclear potential numerically by adaptive quadrature.
+ */
+real_type NuclearZoneBuilder::integrate_potential(real_type rmin,
+                                                  real_type delta_r,
+                                                  real_type integral,
+                                                  real_type ws_shift) const
+{
+    constexpr int max_trials = 1000;
+    constexpr real_type epsilon = 1e-3;
+
+    int depth = 1;
+    real_type interval = delta_r;
+    real_type result = 0;
+
+    bool succeeded = false;
+    int remaining_trials = max_trials;
+    do
+    {
+        delta_r *= this->half();
+
+        real_type r = rmin - delta_r;
+        real_type fi = 0;
+
+        for (int i = 0; i < depth; ++i)
+        {
+            r += interval;
+            fi += (ws_shift > 0) ? r * (r + ws_shift) / (1 + std::exp(r))
+                                 : ipow<2>(r) * std::exp(-ipow<2>(r));
+        }
+
+        result = this->half() * integral + fi * delta_r;
+
+        if (std::fabs(1 - integral / result) < epsilon)
```

Comment:
I think this would be better written with `SoftEqual`? (Use a function-local `constexpr SoftEqual<>{epsilon}`). The SoftEqual also has an absolute tolerance that might be useful (or might be harmful); what's the scale of the expected result?

Reply:
Updated as suggested - at least, good to see that there is no changes test results.